File: csharp/spaceport/Spaceport/SpaceportDepartureBoard.cs
using System;
using System.Collections.Generic;

namespace SpacePort
{
    public class SpaceportDepartureBoard
    {
        private readonly ISpacelineLaunchInfoProvider _provider;
        
        public List<LaunchInfo> LaunchList { get; }
        
        public SpaceportDepartureBoard(ISpacelineLaunchInfoProvider provider)
        {
            _provider  = provider;
            LaunchList = _provider.GetCurrentLaunches();
            SortLaunches();
        }

        private void SortLaunches() {
            //TODO - TEMP TO IMPLEMENT REMOVE TESTS
            //throw new NotImplementedException("sorting logic hasn't been implemented yet");
        }

        public void RemoveInactiveLaunches()
        {
            DateTime now = DateTime.Now;

            for (int i = LaunchList.Count - 1; i >= 0; i-- )
            {
                LaunchInfo launchInfo = LaunchList[i];
                if (launchInfo.Status == LaunchInfo.LaunchStatus.Launched)
                {
                    if (launchInfo.Time.AddMinutes(5).CompareTo(now) <= 0) {
                        LaunchList.RemoveAt(i);
                    }
                } else if (launchInfo.Status == LaunchInfo.LaunchStatus.Scrubbed)
                {
                    if (launchInfo.Time.AddMinutes(10).CompareTo(now) <= 0)
                    {
                        LaunchList.RemoveAt(i);
                    }
                }
            }
        }
        
        public void OnLaunchInfoChanged(LaunchInfo info)
        {            
            if (info.Status == LaunchInfo.LaunchStatus.Delayed)
            {
                HandleDelayedLaunch();
            } else if (info.Status == LaunchInfo.LaunchStatus.Scrubbed)
            {
                HandleScrubbedLaunch();
            }
            else
            {
                AddNewLaunch(info);
            }
        }

        private void AddNewLaunch(LaunchInfo info)
        {
            LaunchList.Add(info);
        }

        private void HandleScrubbedLaunch()
        {
            throw new NotImplementedException();
        }

        private void HandleDelayedLaunch()
        {
            throw new NotImplementedException();
        }

        public void OnCompleted()
        {
            throw new NotImplementedException();
        }

        public void OnError(Exception error)
        {
            throw new NotImplementedException();
        }
    }
}

File: java/spaceport/src/main/java/com/rocketninesolutions/SpaceportDepartureBoard.java
package com.rocketninesolutions;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class SpaceportDepartureBoard {

    private List<LaunchInfo> launchList;

    public SpaceportDepartureBoard(ISpacelineLaunchInfoProvider provider) {
        launchList = provider.getCurrentLaunches();
        sortLaunches();
    }

    public List<LaunchInfo> getLaunchList() {
        return Collections.unmodifiableList(launchList);
    }

    private void sortLaunches() {
        throw new UnsupportedOperationException("TODO - implement the sorting logic!");
    }

    private void addNewLaunch(LaunchInfo launchInfo) {
        launchList.add(launchInfo);
    }

    private void handleDelayedLaunch(LaunchInfo launchInfo) {
        throw new UnsupportedOperationException("delayed launches are not supported");
    }

    private void handleScrubbedLaunch(LaunchInfo launchInfo) {
        throw new UnsupportedOperationException("scrubbed launches are not supported");
    }

    private void handleLaunched(LaunchInfo launchInfo) {
        throw new UnsupportedOperationException("launched launches are not supported");
    }

    protected void removeInactiveLaunches() {
        LocalDateTime now = LocalDateTime.now();

        Iterator<LaunchInfo> iter = launchList.iterator();
        while (iter.hasNext()) {
            LaunchInfo l = iter.next();
            if (LaunchInfo.LaunchStatus.Launched == l.getStatus()) {
                if (l.getTime().plusMinutes(5).compareTo(now) <= 0) {
                    iter.remove();
                }
            } else if (LaunchInfo.LaunchStatus.Scrubbed == l.getStatus()) {
                if (l.getTime().plusMinutes(10).compareTo(now) <= 0) {
                    iter.remove();
                }
            }
        }
    }

    public void onLaunchInfoChanged(LaunchInfo launchInfo) {
        LaunchInfo.LaunchStatus status = launchInfo.getStatus();

        if (status == LaunchInfo.LaunchStatus.Delayed) {
            handleDelayedLaunch(launchInfo);
        } else if (status == LaunchInfo.LaunchStatus.Scrubbed) {
            handleScrubbedLaunch(launchInfo);
        } else if (status == LaunchInfo.LaunchStatus.Launched) {
            handleLaunched(launchInfo);
        } else {
            addNewLaunch(launchInfo);
        }
    }
}
